test/replication-luatest/gh_6033_box_promote_demote_test.lua

```diff
@@ -0,0 +1,505 @@
+        },
+    }
+
+    g.server_1 = g.cluster:build_and_add_server(
+        {alias = 'server_1', engine = engine, box_cfg = g.box_cfg})
+    g.server_2 = g.cluster:build_and_add_server(
+        {alias = 'server_2', engine = engine, box_cfg = g.box_cfg})
+    g.cluster:start()
+end
+
+local cluster_reinit = function(g)
+    g.server_1:wait_lsn(g.server_2)
+    g.server_2:wait_lsn(g.server_1)
+
+    g.server_1:box_config(g.box_cfg)
+    g.server_2:box_config(g.box_cfg)
+end
+
+g_nspaces.before_all(cluster_init)
+g_spaces.before_all(cluster_init)
+
+g_nspaces.after_all(function(g) g.cluster:drop() end)
+g_spaces.after_all(function(g) g.cluster:drop() end)
+
+g_nspaces.after_each(cluster_reinit)
+g_spaces.after_each(cluster_reinit)
+
+-- Promoting/demoting should succeed if server is not configured.
+g_unconfigured.test_unconfigured = function()
+    local ok, err = pcall(box.ctl.promote)
+    luatest.assert(ok, string.format(
+        'error while promoting unconfigured server: %s', err))
+
+    local ok, err = pcall(box.ctl.demote)
+    luatest.assert(ok, string.format(
+        'error while demoting unconfigured server: %s', err))
+end
+
+-- Promoting current raft leader and synchro queue owner should succeed
+-- with elections enabled.
+g_nspaces.test_leader_promote = function(g)
+    g.server_1:box_config({election_mode = 'manual'})
+    g.cluster:promote(g.server_1)
+
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(ok, string.format(
+        'error while promoting leader with elections on: %s', err))
+
+    g.server_1:box_config({election_mode = 'off'})
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(ok, string.format(
+        'error while promoting leader with elections off: %s', err))
+end
+
+g_nspaces.after_test('test_leader_promote', function(g)
+    g.cluster:demote(g.server_1)
+    g.server_2:wait_synchro_queue_owner_id(0)
+end)
+
+-- Demoting current follower should succeed.
+g_nspaces.test_follower_demote = function(g)
+    local ok, err = g.server_2:exec(function()
+        return pcall(box.ctl.demote)
+    end)
+    luatest.assert(ok, string.format(
+        'error while demoting follower with elections off: %s', err))
+
+    g.server_1:box_config({election_mode = 'manual'})
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.demote)
+    end)
+    luatest.assert(ok, string.format(
+        'error while demoting follower with elections on: %s', err))
+end
+
+-- Promoting current raft leader should succeed,
+-- even if he doesn't own synchro queue with elections enabled.
+g_nspaces.test_raft_leader_promote = function(g)
+    g.server_1:box_config({election_mode = 'manual'})
+
+    -- Promote server, but get stuck before obtaining synchro queue
+    -- (write term bump to wal, get stuck on promote)
+    wal_delay_start(g.server_1, 2)
```

Comment:
You here again rely on exact WAL write count when PROMOTE is written. You should better allow WAL writes one by one until you get both WAL_DELAY active + queue being busy. That would mean PROMOTE is currently stuck in WAL. I made a function for my own patch in another ticket which you could use:
```Lua
-- Allow WAL writes one by one until the synchro queue becomes busy. The
-- function is needed, because during box.ctl.promote() it is not known for sure
-- which WAL write is PROMOTE - first, second, third? Even if known, it might
-- change in the future.
local function play_wal_until_synchro_queue_is_busy(server)
    luatest.helpers.retrying({timeout = wait_timeout}, server.exec, server,
                             function()
        if not box.error.injection.get('ERRINJ_WAL_DELAY') then
            error('WAL did not reach the delay yet')
        end
        if box.info.synchro.queue.busy then
            return
        end
        -- Allow 1 more WAL write.
        box.error.injection.set('ERRINJ_WAL_DELAY_COUNTDOWN', 0)
        box.error.injection.set('ERRINJ_WAL_DELAY', false)
        error('Not busy yet')
    end)
end
```
Usage is to make `box.error.injection.set('ERRINJ_WAL_DELAY_COUNTDOWN', 0)`, then `require('fiber').create(box.ctl.promote)`, then `play_wal_until_synchro_queue_is_busy(server)`.

If you don't want to bother, then at least there should be an assertion that `box.info.synchro.queue.busy` is true. Then it probably will fail if PROMOTE WAL write won't be number 2 in the future.